cedar-policy-validator/src/custom_schema/err.rs

```diff
@@ -0,0 +1,322 @@
+use std::{
+    collections::HashMap,
+    fmt::Display,
+    ops::{Deref, DerefMut},
+};
+
+use cedar_policy_core::parser::{err::expected_to_string, unescape::UnescapeError, Loc, Node};
+use lalrpop_util as lalr;
+use lazy_static::lazy_static;
+use miette::{Diagnostic, LabeledSpan, SourceSpan};
+use nonempty::NonEmpty;
+use smol_str::SmolStr;
+use thiserror::Error;
+
+use super::ast::Str;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum UserError {
+    EmptyList,
+    StringEscape(NonEmpty<UnescapeError>),
+}
+
+pub(crate) type RawLocation = usize;
+pub(crate) type RawToken<'a> = lalr::lexer::Token<'a>;
+pub(crate) type RawUserError = Node<UserError>;
+
+pub(crate) type RawParseError<'a> = lalr::ParseError<RawLocation, RawToken<'a>, RawUserError>;
+pub(crate) type RawErrorRecovery<'a> = lalr::ErrorRecovery<RawLocation, RawToken<'a>, RawUserError>;
+
+type OwnedRawParseError = lalr::ParseError<RawLocation, String, RawUserError>;
+
+lazy_static! {
+    /// Keys mirror the token names defined in the `match` block of
+    /// `grammar.lalrpop`.
+    static ref FRIENDLY_TOKEN_NAMES: HashMap<&'static str, &'static str> = HashMap::from([
+        ("IN", "`in`"),
+        ("PRINCIPAL", "`principal`"),
+        ("ACTION", "`action`"),
+        ("RESOURCE", "`resource`"),
+        ("CONTEXT", "`context`"),
+        ("STRINGLIT", "string literal"),
+        ("ENTITY", "`entity`"),
+        ("NAMESPACE", "`namespace`"),
+        ("TYPE", "`type`"),
+    ]);
+}
+
+/// For errors during parsing
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum ParseError {
+    /// Error generated by lalrpop
+    ToAST(OwnedRawParseError),
```

Comment:
Rust style is not to all-caps acronyms in PascalCase identifiers.  For instance, `HttpError` and `JsonError`, not `HTTPError` and `JSONError`.  We aren't super consistent with this in our existing code (except in our public API where we comply), but we might as well follow it in new code.  So this would be `ToAst`